Update tensorflow-probability

ehfd opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * ❌ ``requirements: run_constrained: tensorflow-datasets >= 2.2.0`` should not contain a space between relational operator and the version, i.e. ``tensorflow-datasets >=2.2.0``


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13158962317](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13158962317). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/meta.yaml**:

 * ℹ️ Detected pure Python wheel(s) in source: `https://pypi.org/packages/py2.py3/t/tensorflow-probability/tensorflow_probability-{{ version }}-py2.py3-none-any.whl`. This is generally ok for pure Python wheels and noarch=python packages but it's preferred to use a source distribution (sdist) if possible.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159364433](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159364433). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159558716](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159558716). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159720112](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159720112). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * ❌ The test section contained an unexpected subsection name. command is not a valid subsection name.

For **recipe/meta.yaml**:

 * ℹ️ Detected pure Python wheel(s) in source: `https://pypi.org/packages/py2.py3/t/tensorflow-probability/tensorflow_probability-{{ version }}-py2.py3-none-any.whl`. This is generally ok for pure Python wheels and noarch=python packages but it's preferred to use a source distribution (sdist) if possible.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159874562](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159874562). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159879563](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159879563). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/meta.yaml**:

 * ℹ️ Detected pure Python wheel(s) in source: `https://pypi.org/packages/py2.py3/t/tensorflow-probability/tensorflow_probability-{{ version }}-py2.py3-none-any.whl`. This is generally ok for pure Python wheels and noarch=python packages but it's preferred to use a source distribution (sdist) if possible.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159947731](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159947731). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159962243](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13159962243). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13160118693](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13160118693). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@jonas-eschle I switched a total of three things around, here.

First:

I found that in https://github.com/conda-forge/tensorflow-metadata-feedstock, there is a pure Python wheel file that ships the generated Python files from Bazel from the publishers of the TensorFlow ecosystem.

Therefore, this PR allows maintainers to completely remove Bazel from the workflows, making it more robust and preventing patches, and instead verify that dependencies are correct by checking with `pip check`.

The resulting content should be identical as Bazel merely generates Python files meant to be used anywhere regardless of the environment. I intend to have move all `tensorflow-*` feedstocks that are `noarch` and capable of this approach.

Second:

As per #53, I have moved the optional dependencies to `run_constrained`. Since there will be a message for users when each package is not installed, this is the precise usage case `run_constrained` exists for.

Third:

I have found no justification (aside from your issues https://github.com/tensorflow/probability/issues/1851 and https://github.com/tensorflow/probability/issues/1832) that mandate the tight dependency there is now.
The fact that a release is tested against a certain version of a certain package does not indicate that it is not compatible with other versions.

For example, the current requirement for JAX is very recent and prevents the installation of people using Python 3.9, as well as preventing interoperability with some other package that needs an older version of JAX.
Therefore, I propose that the dependency structure should be set to those exactly equal to those in `setup.py` and not anything tighter.

ehfd commented:
@conda-forge-admin, please rerender

jonas-eschle commented:
Hi @ehfd , thanks a lot for the PR and the detailed explanations!

> First:
> 
> I found that in [conda-forge/tensorflow-metadata-feedstock](https://github.com/conda-forge/tensorflow-metadata-feedstock), there is a pure Python wheel file that ships the generated Python files from Bazel from the publishers of the TensorFlow ecosystem.

Ah, you meant that you're taking the approach of TF metadata basically and want to use this here as well ,relying on the wheel by TFP on pypi instead of building from source? While [conda-forge](https://docs.conda.io/projects/conda-build/en/stable/user-guide/wheel-files.html) seems to prefer, in general, the source builds, I agree that this is a nightmare and the wheels approach seems to fit a lot better here. Agree, thanks a lot!

> Second:
> 
> As per #53, I have moved the optional dependencies to `run_constrained`. Since there will be a message for users when each package is not installed, this is the precise usage case `run_constrained` exists for. However, I retained `tensorflow` and `tf-keras` as mandatory dependencies because `tf-keras` is non-trivial for regular users to discover that they are dependencies while using `tensorflow`.

I agree on moving the deps away, but I would go further and remove tf-keras and (especially) TensorFlow  as well. The main heavy weight is TensorFlow and there is truly no need to have it as a mandatory dependency. Yes, `tf-keras` is a bit of an issue, at the same time, I am not very happy about it's [own maintenance on conda-forge](https://github.com/conda-forge/tf-keras-feedstock/pull/5). This should not hold releases back that do not depend on tf-keras. I've also just opened a [PR](https://github.com/tensorflow/probability/pull/1989) to further improve the error message. Let's just move it to the optional deps as well, do you agree?

> Third:
> 
> I have found no indication that mandate the tight dependency there is now (aside from your issues [tensorflow/probability#1851](https://github.com/tensorflow/probability/issues/1851) and [tensorflow/probability#1832](https://github.com/tensorflow/probability/issues/1832) that nonetheless doesn't imply that the dependency should be tighter than `setup.py` and `required_packages.py`).

There is: `__init__.py` has [a runtime dependency](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/__init__.py#L54) on 2.18. I've opened an issue for clarification: https://github.com/tensorflow/probability/issues/1990

Easier to relax later than to break existing installations. We can move forward with the stricter deps _now_ and maybe relax later with a second build or wait for the response and see.

ehfd commented:
@jonas-eschle Thank you for your insightful feedback. This helps a lot in understanding this feedstock in general.

Number 1: Yep, let's use this approach.
Number 2: I have no objections if you want to take it further than it is now.
Number 3: I have not noticed `__init__.py`. Thank you for pointint this out.

I'll wait for an initial clarification from the project's side on Number 3, and adjust accordingly.

ehfd commented:
Modified 1 and 2.

jonas-eschle commented on recipe/meta.yaml:
```diff
@@ -1,54 +1,56 @@
+    # Required with TensorFlow as this package only supports Keras 2
+    - tf-keras >=2.16
+    - tensorflow-datasets >=2.2.0
 
 test:
+  requires:
```
```suggestion
    - jax >=0.4.35, <0.5.0

test:
  requires:
    - tf-keras
    - tensorflow-base
    - jax
```

Adding JAX to the constraints (relaxed) and tf with keras to the test run requirements

ehfd replied:
Where was the source of this restraint? I don't see this within the code as well.

ehfd replied:
Committed in https://github.com/conda-forge/tensorflow-probability-feedstock/pull/56/commits/e769a74b9a39a2ffe34728e0b752d2154b579b62.

jonas-eschle replied:
Not quite I think. Does 0.4.1 (do you mean 0.4.0) mean the lowest in the 0.4.x series?
If so, JAX [did not](https://docs.jax.dev/en/latest/jep/25516-effver.html#non-semantic-versioning-status-quo) use semantic versioning. There can be breaking changes between 0.4.35 and 0.4.34 (and there are in general). They switched to effortbased versioning. Nice. Still doesn't tell anything about compatibility and therefore doesn't help us here a lot.
I therefore think it's best to keep the lower bound. Or where should we set it? I don't think it's a big deal, as it won't break peoples code. It will only prevent them from upgrading _without upgrading JAX_. But that's it.

For 0.5.0, feel free to test the whole suite, (that would be ideal actualy), if we're reasonably sure that we can lift this restriction, then let's do that. If TFP in pypi doesn't restrict it, why should we here on conda

ehfd replied:
It also seems that JAX 0.5.0 doesn't exist in conda-forge yet, so we don't know.

jonas-eschle commented on recipe/meta.yaml:
```diff
@@ -1,54 +1,56 @@
     - dm-tree
+  run_constrained:
+    # Optional dependencies: tensorflow and tf-keras / tensorflow-datasets / jax and jaxlib (without version constraints)
+    - tensorflow >=2.16.0
```
```suggestion
    - tensorflow-base >=2.16.0
```
TensorFlow has more deps like TensorBoard, so ideally we depend on TensorFlow-base, afaiu, that's the recommended approach

ehfd commented:
> i.e. two weeks

A week perhaps.

jonas-eschle commented:
> > i.e. two weeks
> 
> A week perhaps.

As anticipated, no response. But alas, the issue is actually simpler than thought: if TFP is run with a TF version <2.18, it breaks on runtime. Therefore, we need to require 2.18.
If they relax it in the future, we can relax it too oc.

ehfd commented:
@jonas-eschle 

I've tightened the version up, waiting on https://github.com/conda-forge/tf-keras-feedstock/pull/8.
The build now will fail because of the unavailability.

Two more things, here.

1. Does the previous build (v0.24.0 - stable against TensorFlow 2.16.1 and JAX 0.4.25) work with Tensorflow 2.17.x or 2.18.x?
2. Is the JAX version dependency correct?

jonas-eschle commented:
> @jonas-eschle
> 
> I've tightened the version pinnings up, waiting on [conda-forge/tf-keras-feedstock#8](https://github.com/conda-forge/tf-keras-feedstock/pull/8). The build now will fail because of the unavailability.

Agree, seems all fine
 
> 1. Does the previous build (v0.24.0 - stable against TensorFlow 2.16.1 and JAX 0.4.25) work with Tensorflow 2.17.x and/or 2.18.x or did it not (in order to decide `>=2.18.0,<2.19.0` or `>=2.18.0`?
I think it "works", however, it's also likely that it will fail with newer versions. If we leave it open, in general, the newest version will be picked of TFP, but if a lower TFP version is required, maybe a newer TF version which actually breaks things will be installed. We could still restrict things then (or restrict now and release later). As TFP and TF get more stable with every release, I think I am now comfortable to _not_ restrict the upper version. But it's a choice, open for both.

> 2. Is the JAX version dependency adequate (as in [Update tensorflow-probability #56 (comment)](https://github.com/conda-forge/tensorflow-probability-feedstock/pull/56#discussion_r1956674270))?

In short: lower restrictions seem fine as they are, upper _can_ be removed for both JAX (if tested) and TF. No strong feelings, we can always patch things. Let me know what you think, adjust if you like, and then we're good to go IMHO

ehfd commented:
@jonas-eschle Need to test the previous tensorflow-probability version on 2.18, so let's wait until tf-keras 2.18 is added.

ehfd commented:
The error message for people in the future:

`ImportError: This version of TensorFlow Probability requires TensorFlow version >= 2.18; Detected an installation of version 2.17.0. Please upgrade TensorFlow to proceed.`

jonas-eschle commented:
> @jonas-eschle
> 
> So conclusion: This version depends on tf-keras >=2.18 so this can only be merged when that is merged. TensorFlow has explicit import errors when the version is lower than designated (2.18.0 for 0.25.0). JAX version requirements should NOT be lowered, as it breaks at any time.

Agree. We just don't know about JAX, but I don't think it's a disadvantage, really. If you wat to upgrade, why not also upgrade JAX? And if there is a good reason not to, i.e. because JAX had breaking changes, chances are, that _exactly these_ will also break TFP. So yes, agree.

> For whether the upward bound dependency constraints can be relieved, need to test tensorflow-probability 0.24.0 with TensorFlow 2.18.0, and test JAX 0.5.x on tensorflow-probability 0.24.0 and 0.25.0. Both need to wait for respective feedstocks.

Agree. I would suggest not to make these "blocking" items for now: as soon as TF-Keras is out, we should release this. Preferrably with the restrictions in place (the jump to 0.5 could indicate something larger, so maybe we make one at a time/resubmit an increased boundary/no boundary if we're confident. Always ugly if code breaks at runtime error and you need to figure out that this is due to a version mismatch, i.e. too new JAX version for example). Much harder than finding a way around the dependency.

ehfd commented:
> Agree. We just don't know about JAX, but I don't think it's a disadvantage, really. If you wat to upgrade, why not also upgrade JAX? And if there is a good reason not to, i.e. because JAX had breaking changes, chances are, that exactly these will also break TFP. So yes, agree.

"conda-forge JAX maintainer" sounds nice, I'll try.

jonas-eschle commented:
> It seems that something crazy is going on to make jaxlib ([conda-forge/jaxlib-feedstock#294 (files)](https://github.com/conda-forge/jaxlib-feedstock/pull/294/files)) work and I don't think I'm really ready for this yet.
> 
> I still don't understand Bazel well enough and the peripheral TensorFlow packages (mostly tensorflow-datasets, array-record, tensorflow-text, etc.) were supposed to be an opportunity to get deeper into this before being introduced into the more important packages.

Same here, my knowledge of Bazel is very limited. But it seems as if it works, so let's see! (btw, if you wanna just test, you can use the pip version, right?)

ehfd commented:
> if you wanna just test, you can use the pip version, right?

You are right, I just made the same assumptions as TensorFlow where you need to use a Conda build to verify CUDA, when the topic here isn't CUDA.

ehfd commented:
@jonas-eschle The combination of TensorFlow 2.18.0 (conda-forge) and tf-keras 2.18.0 (pip) with TensorFlow Probability 0.24.0 did not present any issues.

[test_tf.tar.gz](https://github.com/user-attachments/files/18885610/test_tf.tar.gz)

```
E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1740050377.882254   34256 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1740050377.889672   34256 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-02-20 11:19:37.909118: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
TensorFlow version: 2.18.0
TensorFlow Probability version: 0.24.0
I0000 00:00:1740050381.406957   34256 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 15212 MB memory:  -> device: 0, name: Quadro RTX 6000, pci bus id: 0000:3b:00.0, compute capability: 7.5

[Basic Distribution]
Sample mean = 0.011089784
Average log probability = -1.4298217

[Joint Distribution]
Joint sample (x, y): [np.float32(0.58516246), np.float32(1.5285759)]

[Bayesian Linear Regression via HMC]
Posterior mean estimate of a: 1.207
Posterior mean estimate of b: 0.742
Posterior mean estimate of sigma: 0.011
Matplotlib not installed; skipping plots.
```

```
E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1740050411.463516   34979 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1740050411.471401   34979 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-02-20 11:20:11.497504: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1740050414.886086   34979 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 15212 MB memory:  -> device: 0, name: Quadro RTX 6000, pci bus id: 0000:3b:00.0, compute capability: 7.5
Normal Distribution Mean: 0.0
Normal Distribution Variance: 1.0
Sample from Normal: [ 0.73769367  1.1850613  -2.313611    0.2876023   0.51588637]
Log Probabilities: [-0.9189385 -1.4189385 -1.4189385]
CDF Values: [0.5        0.8413447  0.15865527]
Lognormal Sample: [2.7763894  0.54644555 0.6659182  2.9804587  2.1727817 ]
Joint Distribution Sample: [<tf.Tensor: shape=(), dtype=float32, numpy=1.496046543121338>, <tf.Tensor: shape=(), dtype=float32, numpy=0.6264479160308838>]
Log Probability of Joint Sample: -2.7930865
Estimated Mean: -0.042495437
Estimated Variance: 0.9843308
TensorFlow Probability tests completed successfully.
```

ehfd commented:
@conda-forge-admin, please rerender

jonas-eschle commented:
> @jonas-eschle The combination of TensorFlow 2.18.0 (conda-forge) and tf-keras 2.18.0 (pip) with TensorFlow Probability 0.24.0 (the older version) did not present any issues.

Yes, however, this is just testing a tiny subset of the functionality. If we want to assure that things work, we would need to run _at least_ the whole test suite. And ideally on different machines, i.e. Linux, OSX, Windows....
That's by far not enough to know by now.

Also, just because the old one did not break does not mean that newer version won't break.

So I would still say for JAX, if they do a major upgrade to 0.5, it's save to not yet release it. For TF, I am less sure what (and if) to constrain it.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13434217828](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13434217828). Examine the logs at this URL for more detail.</sub>

ehfd commented:
> So I would still say for JAX, if they do a major upgrade to 0.5, it's save to not yet release it.

I will reverting the constraints for JAX.
However, the issue here is the current constraints restrict available packages for TensorFlow 2.17.x and potentially 2.19.x.

jonas-eschle commented:
> However, the issue here is the current constraints restrict available packages for TensorFlow 2.17.x

True. But this isn't our choice, right? TFP simply _won't_ work wit a lower version, it's in the code

> It is suspected that `pip` users are fine with loosened constraints.

Actually, good argument. Pip has the same problem. Why be more constraint here. I am fine with removing the upper constraint for JAX as well, the argument convinced me now

ehfd commented:
> However, the issue here is the current constraints restrict available packages for TensorFlow 2.17.x
> True. But this isn't our choice, right? TFP simply won't work wit a lower version, it's in the code

The argument is not with the past, but the future (2.19.x potentially).

> Actually, good argument. Pip has the same problem. Why be more constraint here. I am fine with removing the upper constraint for JAX as well, the argument convinced me now

Sure, I will remove the upper bound constraints again.

An older TensorFlow version then 2.18.0 is blocked, obviously, and key capabilities in JAX breaks with TensorFlow Probability 0.25.0 at around version 0.4.24 or 0.4.23 (I have tested so). So decreasing the lower limits is risky.

ehfd commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13435906324](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13435906324). Examine the logs at this URL for more detail.</sub>

ehfd commented:
@conda-forge-admin, please rerender

ehfd commented:
@jonas-eschle And the builds succeed...

I think if there is an issue with the upper bounds somehow, we can limit it when there is an issue.

jonas-eschle commented:
> An older TensorFlow version then 2.18.0 is blocked, obviously, and key capabilities in JAX breaks with TensorFlow Probability 0.25.0 at around version 0.4.24 or 0.4.23 (I have tested so). So decreasing the lower limits is risky, and should be kept, whether or not pip enforces it. But the upper limits - I doubt there will be issues.

Very well, I may misunderstood, but we agree.

> I think if there is an issue with the upper bounds somehow, we can limit it when there is an issue, since pip would under the same problem.

It's not that easy, and the argument also works the other way around: you can relax it once we know it works. I am saying that it's not that easy, because how do you know it broke because of a mismatched version? A user may randomly bumps into it. But that's not a version error. That's a random error. And the user tracks it down, may first thinks it's him. If he is a power user, he can track it down and, fair enough, report it _maybe_ in TFP (or here). But there are a lot of ifs. Most notably, if will cause _hard to understand_ bugs for users.


But we agree, LGTM! Many thanks for this contribution!

ehfd commented:
Yeah, learned a lot of things in the progress. Thanks for your continued patience. @jonas-eschle
